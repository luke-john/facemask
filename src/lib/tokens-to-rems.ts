import * as rem from "polished/lib/helpers/rem";

export interface PxTokens {
  [scale: string]: {
    [key: string]: number[];
  };
}

export interface RemTokens {
  [scale: string]: {
    [key: string]: string[];
  };
}

export const tokensToRems = <UsedTokens extends PxTokens>(
  tokens: UsedTokens,
  customFontSizes?: number[]
): UsedTokens => {
  const convertTokenSet = (tokenSet: { [key: string]: number[] }) => {
    const convertedTokenSet = {};
    const keys = Object.keys(tokenSet) as Array<keyof typeof tokenSet>;

    keys.forEach(tokenKey => {
      convertedTokenSet[tokenKey] = tokenSet[tokenKey].map(
        (tokenValue, index) => {
          const baseFontSize = customFontSizes ? customFontSizes[index] : 16;

          return rem(tokenValue, baseFontSize);
        }
      );
    });

    return convertedTokenSet;
  };

  const remTokens: RemTokens = {};

  Object.keys(tokens).forEach(scaleKey => {
    remTokens[scaleKey] = convertTokenSet(tokens[scaleKey]);
  });

  // The two typings are both acceptable by styled-components,
  // we use UsedTokens for better type hinting/developer experience
  // tslint:disable-next-line:no-any
  return (remTokens as any) as UsedTokens;
};
