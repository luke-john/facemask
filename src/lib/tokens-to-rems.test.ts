import { tokensToRems } from "./tokens-to-rems";

const exampleTokens = {
  spacing: {
    moon: [4, 6, 10],
    mars: [8, 13, 20],
    earth: [16, 26, 41],
    saturn: [32, 52, 83],
    sun: [66, 106, 168]
  },
  type: {
    snail: [13, 16, 20],
    mouse: [16, 20, 26],
    otter: [20, 26, 32],
    dolphin: [26, 32, 41],
    hippo: [32, 41, 52],
    mammoth: [41, 52, 66],
    whale: [52, 66, 83]
  }
};

it("converts pixels to rems", () => {
  expect(tokensToRems(exampleTokens)).toMatchSnapshot();
});

it("converts pixels to rems with custom base font sizes", () => {
  expect(tokensToRems(exampleTokens, [10, 15, 20])).toMatchSnapshot();
});
