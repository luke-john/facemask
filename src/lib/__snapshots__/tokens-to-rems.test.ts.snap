// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts pixels to rems 1`] = `
Object {
  "spacing": Object {
    "earth": Array [
      "1rem",
      "1.625rem",
      "2.5625rem",
    ],
    "mars": Array [
      "0.5rem",
      "0.8125rem",
      "1.25rem",
    ],
    "moon": Array [
      "0.25rem",
      "0.375rem",
      "0.625rem",
    ],
    "saturn": Array [
      "2rem",
      "3.25rem",
      "5.1875rem",
    ],
    "sun": Array [
      "4.125rem",
      "6.625rem",
      "10.5rem",
    ],
  },
  "type": Object {
    "dolphin": Array [
      "1.625rem",
      "2rem",
      "2.5625rem",
    ],
    "hippo": Array [
      "2rem",
      "2.5625rem",
      "3.25rem",
    ],
    "mammoth": Array [
      "2.5625rem",
      "3.25rem",
      "4.125rem",
    ],
    "mouse": Array [
      "1rem",
      "1.25rem",
      "1.625rem",
    ],
    "otter": Array [
      "1.25rem",
      "1.625rem",
      "2rem",
    ],
    "snail": Array [
      "0.8125rem",
      "1rem",
      "1.25rem",
    ],
    "whale": Array [
      "3.25rem",
      "4.125rem",
      "5.1875rem",
    ],
  },
}
`;

exports[`converts pixels to rems with custom base font sizes 1`] = `
Object {
  "spacing": Object {
    "earth": Array [
      "1.6rem",
      "1.7333333333333334rem",
      "2.05rem",
    ],
    "mars": Array [
      "0.8rem",
      "0.8666666666666667rem",
      "1rem",
    ],
    "moon": Array [
      "0.4rem",
      "0.4rem",
      "0.5rem",
    ],
    "saturn": Array [
      "3.2rem",
      "3.466666666666667rem",
      "4.15rem",
    ],
    "sun": Array [
      "6.6rem",
      "7.066666666666666rem",
      "8.4rem",
    ],
  },
  "type": Object {
    "dolphin": Array [
      "2.6rem",
      "2.1333333333333333rem",
      "2.05rem",
    ],
    "hippo": Array [
      "3.2rem",
      "2.7333333333333334rem",
      "2.6rem",
    ],
    "mammoth": Array [
      "4.1rem",
      "3.466666666666667rem",
      "3.3rem",
    ],
    "mouse": Array [
      "1.6rem",
      "1.3333333333333333rem",
      "1.3rem",
    ],
    "otter": Array [
      "2rem",
      "1.7333333333333334rem",
      "1.6rem",
    ],
    "snail": Array [
      "1.3rem",
      "1.0666666666666667rem",
      "1rem",
    ],
    "whale": Array [
      "5.2rem",
      "4.4rem",
      "4.15rem",
    ],
  },
}
`;
